### PR TITLE
[MoE] Alpha MoE integration

### PR DESCRIPTION
**Pull request by vyom-hai**
## Purpose

Add support for [Alpha-MoE](https://github.com/NVIDIA/Alpha-MoE), a high-performance fused Mixture-of-Experts CUDA megakernel optimized for tensor-parallel servings of FP8 quantized MoE models.

This integration provides an alternative MoE kernel backend that fuses the up projection, SiLU activation, and down projection into a single CUDA kernel, reducing memory bandwidth overhead and improving inference latency for models like DeepSeek-V3.

**Changes:**
- **New file: `vllm/model_executor/layers/fused_moe/alpha_moe.py`** - Alpha-MoE kernel wrapper with availability check, weight interleaving, and kernel dispatch
- **Modified: `vllm/model_executor/layers/fused_moe/fused_moe.py`** - Added Alpha-MoE to the kernel dispatch chain (before DeepGemm fallback)
- **Modified: `vllm/model_executor/layers/fused_moe/layer.py`** - Added `_maybe_interleave_for_alpha_moe()` for automatic weight interleaving during model loading
- **Modified: `vllm/envs.py`** - Added `VLLM_USE_ALPHA_MOE` and `VLLM_ALPHA_MOE_CONFIG` environment variables

### Clone Alpha-MoE repository
git clone https://github.com/Aleph-Alpha/Alpha-MoE.git
cd Alpha-MoE

### Install with CUDA extension compilation (requires CUDA toolkit)
pip install -e . --no-build-isolation

### Set library path for PyTorch dependencies
export LD_LIBRARY_PATH="$(python -c 'import torch; print(torch.__path__[0])')/lib:$LD_LIBRARY_PATH"

### Generate JIT kernel configuration (optional but recommended)

cd Alpha-MoE
python jit_moe.py --E 256 --N 2048 --K 7168 --top-k 8 --out-file configs/moe_jit_config.json
### Run latency benchmark with Alpha-MoE

VLLM_USE_ALPHA_MOE=1 \
VLLM_ALPHA_MOE_CONFIG=<path-to-alpha-moe>/configs/moe_jit_config.json \
vllm bench latency \
  --model deepseek-ai/DeepSeek-V3 \
  --tensor-parallel-size 8 \
  --quantization fp8 \
  --enforce-eager \
  --num-iters 8 \
  --batch-size 16 \
  --max-model-len 16384

## Test Result
**Model:** DeepSeek-V3 | **Hardware:** 8x NVIDIA H200 GPUs | **Quantization:** FP8 | **Batch Size:** 16

| Metric | Alpha-MoE | DeepGemm (Baseline) | Improvement |
|--------|-----------|---------------------|-------------|
| **Avg Latency** | 10.916s | 12.999s | **16.0% faster** |
| P10 Latency | 10.839s | 12.846s | 15.6% faster |
| P25 Latency | 10.852s | 12.863s | 15.6% faster |
| P50 Latency | 10.875s | 12.901s | 15.7% faster |
| P75 Latency | 10.885s | 12.922s | 15.8% faster |
| P90 Latency | 11.006s | 13.189s | 16.5% faster |
| P99 Latency | 11.251s | 13.715s | **18.0% faster** |
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> Introduces an optional high-performance MoE backend using Alpha‑MoE FP8 fused kernels, enabled via env vars and selected ahead of other backends when valid.
> 
> - New `vllm/model_executor/layers/fused_moe/alpha_moe.py`: availability checks, config loading (`VLLM_ALPHA_MOE_CONFIG`), tensor interleaving for `w1`/gate, and fused `alpha_moe_fused_experts` dispatch
> - `fused_moe.py`: adds Alpha‑MoE to `fused_experts` dispatch path (pre‑DeepGemm) via `_valid_alpha_moe` and `alpha_moe_fused_experts`
> - `layer.py`: implements `_maybe_interleave_for_alpha_moe` to auto‑interleave `w13` weights/scales after both halves load; tracks per‑expert chunks
> - `envs.py`: adds `VLLM_USE_ALPHA_MOE` and `VLLM_ALPHA_MOE_CONFIG` with wiring into `environment_variables`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8ae7ba4d4db45492dd5844e3f9f0edf81ad9e4d8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds an optional high‑performance Alpha‑MoE FP8 fused kernel path and integrates it into MoE execution.
> 
> - New `vllm/model_executor/layers/fused_moe/alpha_moe.py`: availability/env checks, JSON config loading (`VLLM_ALPHA_MOE_CONFIG`), `interleave_tensor` for `w1`/gate, `_valid_alpha_moe`, and fused `alpha_moe_fused_experts` kernel call
> - `fused_moe.py`: routes `fused_experts` to Alpha‑MoE when `_valid_alpha_moe` passes (before DeepGemm/other paths)
> - `layer.py`: `_maybe_interleave_for_alpha_moe` to auto‑interleave `w13` weights/scales after both halves load; tracks per‑expert chunks; gated by `is_alpha_moe_enabled`
> - `envs.py`: adds `VLLM_USE_ALPHA_MOE` and `VLLM_ALPHA_MOE_CONFIG` environment variables and wires them into `environment_variables`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 448112e3bb5d78162df37abc666649cc4489142a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds an optional high-performance MoE backend using Alpha‑MoE FP8 fused kernels, enabled via `VLLM_USE_ALPHA_MOE` with optional `VLLM_ALPHA_MOE_CONFIG`.
> 
> - New `vllm/model_executor/layers/fused_moe/alpha_moe.py`: availability checks, JSON config loading/caching, `interleave_tensor` for w1/gate, runtime validation, and fused `alpha_moe_fused_experts` execution (FP8 input quant, token alignment, single-kernel up+SiLU+down)
> - Update `fused_moe.py`: integrate Alpha‑MoE into `fused_experts` dispatch (checked via `_valid_alpha_moe`), taking precedence over DeepGemm when applicable
> - Update `layer.py`: `_maybe_interleave_for_alpha_moe` to auto‑interleave `w13` weights and weight scales after both halves load; tracks per‑expert chunks and guards by activation/router/expert_map conditions
> - Update `envs.py`: add `VLLM_USE_ALPHA_MOE` and `VLLM_ALPHA_MOE_CONFIG` environment variables
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 448112e3bb5d78162df37abc666649cc4489142a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by SzymonOzog** [comment]
Amazing work! I'll look at the code later when I have a bit more time.
Curious how are you running it on B200? The code uses WGMMA and AFAIK it
should not compile on Blackwell but I never tried it

File: vllm/model_executor/layers/fused_moe/alpha_moe.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        - block_n: Block size in N dimension
+        - warp_n: Number of warps in N dimension
+        - stages: Pipeline stages
+    """
+    if config_path is None:
+        config_path = envs.VLLM_ALPHA_MOE_CONFIG
+
+    if config_path is None:
+        # Return default configuration if no config file specified
+        return {"block_m": 128, "block_n": 256, "warp_n": 4, "stages": 3}
```
Comment (SzymonOzog) [inline vllm/model_executor/layers/fused_moe/alpha_moe.py]: The default config in here would throw an assertion error https://github.com/Aleph-Alpha/Alpha-MoE/blob/8960860fde1ea4b7e47a21ee9ecb7d3fae25ddc0/csrc/torch_interface.cpp#L62

File: vllm/model_executor/layers/fused_moe/alpha_moe.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    dist = float("inf")
+    ret = None
+    for nt, val in best_conf.items():
+        if abs(int(nt) - num_tokens) < dist:
+            dist = abs(int(nt) - num_tokens)
+            ret = val
+
+    if ret is None:
+        # Fallback to default if config is empty
+        return {"block_m": 128, "block_n": 256, "warp_n": 4, "stages": 3}
```
Comment (SzymonOzog) [inline vllm/model_executor/layers/fused_moe/alpha_moe.py]: Same issue as above, this would throw an assertion error

**Review by SzymonOzog — CHANGES_REQUESTED** [review]

**Comment by vyom-hai** [comment]
@SzymonOzog Sorry it was compiled on H200s, I mistakenly put B200 in the results section.
